src/qiboml/models/keras.py

```diff
@@ -0,0 +1,65 @@
+"""Keras interface to qiboml layers"""
+
+from dataclasses import dataclass
+
+import keras
+import numpy as np
+import tensorflow as tf  # pylint: disable=import-error
+from qibo.config import raise_error
+
+import qiboml.models.encoding_decoding as ed
+from qiboml.models.abstract import QuantumCircuitLayer
+
+
+@dataclass
+class QuantumModel(keras.layers.Layer):  # pylint: disable=no-member
```

Comment:
Let's say that it is more in a ML style, and indeed as you are saying you only need encoding, circuit and decoding, that you could see as three different layers stacked on top of each other. The only difference is that here you have the flexibility to construct also the circuit modularly, seen as different layers.

Regarding the encoding, I am not sure it's fully classical, in the sense that the input is usually classical but the output should be quantum in general. Instead, the decoding could be an observable, yes, but I would like to offer the possibility of returning other things, like for examples the probabilities or the samples.

In any case the `QuantumLayer` you are suggesting is implemented here already, it's just called in a different way. Namely, the `keras` interface `qiboml.models.keras.QuantumModel` with an `ExpectationLayer` as decoding is precisely taking care of that.

Reply:
> The only difference is that here you have the flexibility to construct also the circuit modularly, seen as different layers.

But this you already have: different circuits are already "layers" that you can concatenate.
Simply, there are red layers and blue layers (quantum and classical) that doesn't make sense to mix. So, to me, this is better implemented if you don't provide (and then possibly deny) the option of concatenating them. But simply keeping them as different and non-compatible objects.

Reply:
> Regarding the encoding, I am not sure it's fully classical, in the sense that the input is usually classical but the output should be quantum in general.

In which sense the output is quantum? Usually you encode in parameters.

If instead you're encoding in a state, this is already cutting Qiboml out of hardware support. Much better than to have a function that is emitting a circuit (thus homogeneous to the *circuit* part) rather than a tensor.

My take on the goal of Qiboml is more in the direction of making Qibo compatible with machine learning, not to have an ML-practitioner-friendly interface to Qibo (since Qibo already has an interface, and doesn't need a wrapper).

Reply:
> Consider for example the binary or phase encoding, you have some classical inputs and you convert them in a layer of X or rotation gates. Of course the gate are represented by the index of the qubit they act on or the parameter, which are classical data, but intrinsically the encoding takes classical data and embeds it in a quantum circuit. Thus I wouldn't call it fully classical.

I would classify the X as part of the circuit, and the indexing as part of the encoding. You can not make a *previous* layer that is stacked before the circuit, since your encoding might also be in the middle of the circuit. In that case, it doesn't make sense to see the rest of the circuit as a separate layer from the encoding.